File: deck.py
import random

class Deck:
    """
    Deck class that manages a standard 52-card deck for Texas Hold'em Poker.
    Supports shuffling, drawing, and tracking remaining cards.
    """

    def __init__(self):
        """
        Initializes a standard 52-card deck.
        Suits: Spades (♠), Hearts (♥), Diamonds (♦), Clubs (♣).
        Ranks: 2-10, Jack, Queen, King, Ace.
        """
        self.suits = ['♠', '♥', '♦', '♣']
        self.ranks = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A']
        self.cards = [rank + suit for suit in self.suits for rank in self.ranks]
        self.shuffle()

    def shuffle(self):
        """
        Shuffles the deck randomly.
        """
        random.shuffle(self.cards)

    def draw(self, count=1):
        """
        Draws the specified number of cards from the deck.
        Args:
            count (int): Number of cards to draw.
        Returns:
            list: A list of drawn cards.
        """
        if count > len(self.cards):
            raise ValueError("Not enough cards remaining in the deck.")
        drawn_cards = self.cards[:count]
        self.cards = self.cards[count:]
        return drawn_cards

    def remaining(self):
        """
        Returns the number of remaining cards in the deck.
        """
        return len(self.cards)

File: deck.c
#include "deck.h"

// Initializes the deck with 52 standard playing cards
void initialize_deck(Deck *deck) {
    char suits[] = {'S', 'H', 'D', 'C'}; // Spades, Hearts, Diamonds, Clubs
    char ranks[] = {'2', '3', '4', '5', '6', '7', '8', '9', 'T', 'J', 'Q', 'K', 'A'};
    int index = 0;

    for (int s = 0; s < 4; s++) {
        for (int r = 0; r < 13; r++) {
            deck->cards[index].suit = suits[s];
            deck->cards[index].rank = ranks[r];
            index++;
        }
    }
    deck->top_card_index = 0;
}

// Shuffles the deck using Fisher-Yates algorithm
void shuffle_deck(Deck *deck) {
    srand(time(NULL));
    for (int i = DECK_SIZE - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        Card temp = deck->cards[i];
        deck->cards[i] = deck->cards[j];
        deck->cards[j] = temp;
    }
}

// Draws a card from the top of the deck
Card draw_card(Deck *deck) {
    if (deck->top_card_index >= DECK_SIZE) {
        fprintf(stderr, "Deck is empty. Cannot draw.");
        exit(EXIT_FAILURE);
    }
    return deck->cards[deck->top_card_index++];
}
